source/LiteralModel.cpp

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include <mariana-trench/EventLogger.h>
+#include <mariana-trench/LiteralModel.h>
+#include <mariana-trench/Positions.h>
+
+namespace marianatrench {
+
+namespace {
+
+class LiteralModelConsistencyError {
+ public:
+  static void raise(const std::string& what) {
+    ERROR(1, "Literal Model Consistency Error: {}", what);
+    EventLogger::log_event("regex_model_consistency_error", what);
+  }
+};
+
+}
+
+static void check_taint_config_consistency(const std::string& pattern, const TaintConfig& frame) {
+  if (frame.kind() == nullptr) {
+    LiteralModelConsistencyError::raise(fmt::format(
+        "Model for regex `{}` must have a source kind.", pattern));
+  }
+}
+
+static Taint create_taint(const std::string& pattern, const std::vector<TaintConfig>& sources) {
+  Taint all_sources;
+  for (const auto& source : sources) {
+    mt_assert(source.is_leaf());
```

Comment:
If this assert fails, wouldn't it be a user error? (i.e. json with callee specified)

would be nicer to raise a JsonValidationError

Reply:
Yes, I think this would be a user error. I used `mt_assert` because that's what's used in `FieldModel` in a similar context:

```C++
void FieldModel::add_source(TaintConfig source) {
  mt_assert(source.is_leaf());
  check_taint_config_consistency(source, "source");
  add_source(Taint{std::move(source)});
}
```

I now added an additional check in `LiteralModel::from_json` where we have the JSON value for context:

```C++
if (!taintConfig.is_leaf())
  throw JsonValidationError{source, /* field */ "sources", "empty callee"};
```

So if we still violate the `mt_assert` in `create_taint`, it would be an application logic error.